src/engine.rs

```diff
@@ -2230,4 +2285,128 @@ mod tests {
             assert!(engine.file_span(LogQueue::Append).0 > start);
         }
     }
+
+    #[test]
+    fn test_start_engine_without_recycle() {
+        let dir = tempfile::Builder::new()
+            .prefix("test_start_without_recycle")
+            .tempdir()
+            .unwrap();
+        let path = dir.path().to_str().unwrap();
+        let cfg = Config {
+            dir: path.to_owned(),
+            enable_log_recycle: false,
+            ..Default::default()
+        };
+        let file_system = Arc::new(DeleteMonitoredFileSystem::new());
+        let engine =
+            RaftLogEngine::open_with_file_system(cfg.clone(), file_system.clone()).unwrap();
+        let (start, _) = engine.file_span(LogQueue::Append);
+        // Only one valid file left, the last one => active_file.
+        assert_eq!(engine.file_count(Some(LogQueue::Append)), 1);
+        assert_eq!(file_system.inner.file_count(), engine.file_count(None));
+        // Append data.
+        let entry_data = vec![b'x'; 512];
+        for rid in 1..=5 {
+            engine.append(rid, 11, 20, Some(&entry_data));
+        }
+        assert_eq!(engine.file_span(LogQueue::Append).0, start);
+        let (start, end) = engine.file_span(LogQueue::Append);
+        assert_eq!(file_system.inner.file_count(), engine.file_count(None));
+        drop(engine);
+        // Reopen the engine with a smaller capacity.
+        let cfg_v2 = Config {
+            target_file_size: ReadableSize::kb(2),
+            purge_threshold: ReadableSize::kb(32),
+            enable_log_recycle: true,
+            prefill_for_recycle: true,
+            ..cfg
+        };
+        let engine = RaftLogEngine::open_with_file_system(cfg_v2, file_system.clone()).unwrap();
+        assert_eq!(engine.file_span(LogQueue::Append), (start, end));
+        // As the recycling open, it would generate stale logs for recycling.
+        let stale_count = file_system.inner.file_count() - engine.file_count(None);
+        assert!(stale_count > 0);
+        // Stale files haven't filled the LogQueue::Append, purge_expired_files won't
+        // make difference.
+        engine.purge_expired_files().unwrap();
+        assert_eq!(engine.file_span(LogQueue::Append), (start, end));
+        for rid in 1..=20 {
+            engine.append(rid, 20, 31, Some(&entry_data));
+        }
+        engine.purge_expired_files().unwrap();
+        assert!(engine.file_span(LogQueue::Append).0 > start);
+        let (start, end) = engine.file_span(LogQueue::Append);
+        let engine = engine.reopen();
+        assert_eq!(engine.file_span(LogQueue::Append), (start, end));
+        // Stale files will be reused for new append files.
+        assert!(stale_count > (file_system.inner.file_count() - engine.file_count(None)));
```

Comment:
Move this check right after the `append`, and check the `stale_count == capacity` here.